sequana/databases.py

```diff
@@ -270,47 +189,277 @@ def switch_header_to_gi(self, acc):
 
-    def download_phage(self):
-        self.download_fasta("phage.txt", "Phages")
+        self._df_nodes_taxon = self.df_nodes.copy()
+        self._df_nodes_taxon.set_index('taxon', inplace=True)
 
-    def download_archaealvirus(self):
-        self.download_fasta("archaealvirus.txt", "ArchaeaViruses")
+    def download_taxdump(self, outpath="."): #pragma: no cover
+        
+        FTP = "ftp.ncbi.nih.gov"
+        execute(f"wget {FTP}/pub/taxonomy/taxdump.tar.gz --directory-prefix {outpath}")
+        execute(f'tar xvfz {outpath}/taxdump.tar.gz -C {outpath}')
 
-    def download_archaea(self):
-        self.download_fasta(*self._metadata["archaea"])
+    def get_number_taxon(self):
+        """Return number of unique taxon"""
+        return len(self.df_names['taxon'].unique())
 
-    def download_bacteria(self):
-        """ organisms (may 2016)
+    def get_average_name_per_taxon(self):
+        """Return number of rows/names per node/taxon"""
+        from sequana.lazy import numpy as np
+        return np.mean([len(values) for values in self._group_name.values()])
 
-        .. note:: this download method is the longest to end. It took about 20mins on 
-            a good connection.
+    def get_scientific_name(self, taxon):
+        """Return scientific name of a given Taxon"""
+        # Takes 2 minutes to scan all taxons
+        return self._subdf.loc[taxon].values[0]
+
+    def get_taxon_from_scientific_name(self, scname):
+        """Return taxon corresponding to a scientific name
+
+        return: unique taxon or first one found. If none found, returns None
+        """
+        res = self.df_names.query("@scname in name")['taxon']
+        return res
+
+    def search(self, name):
+        """Search names colum"""
+        return self.df_names[self.df_names['name'].apply(lambda x : name in x)]
+
+    def get_family(self, taxon):
+        """Get all parent taxons"""
+        taxons = [1]
+        df = self._df_nodes_taxon
+        while True:
+            res = df.loc[taxon]
+            taxons.append(taxon)
+            taxon = res['parent_taxon']
+            # hopefully there is always a family link to 0 or 1
+            if len(res) == 0 or taxon in [0,1]:
+                break
+        return taxons
+
+    def filter_nodes_dmp_file(self, output="nodes_filtered.dmp", taxons=[]):
+        """Save a subset of nodes.dmp given list of valid taxons
+
+        :param str:  Defaults to "nodes_filtered.dmp".
+        :param list taxons:
         """
-        self.download_fasta("bacteria.txt", "Bacteria")
+        all_taxons = set()
+        for i, taxon in enumerate(taxons):
+            parents = self.get_family(taxon)
+            all_taxons.update(parents)
+        N = len(all_taxons)
+        print(f"Saving {N} taxons") 
+        print(all_taxons)
+
+        # In theory one line per taxon
+        with open(self.filename_nodes, "r") as fin:
+            with open(output, "w") as fout:
```

Comment:
Tu peux mettre tes 2 `open` dans le même `with`

Reply:
okay noté. question d'habitude ici. black met a jour ce genre de syntaxe ?

Reply:
Black ne met pas a jour ce genre de syntaxe comme c'est une syntaxe valide 😄